### ISSUE TITLE
SAC : why actor has a target network? Why ModSAC has a Reliable lamdba and TTUR?

### ISSUE THREAD

**Issue by wsgdrfz** [AI4Finance-LLC/ElegantRL#14]
你好 我看到在代码中，sac的actor也有target_net。这个在其他implementation，比如stable_baseline3, spinning_up都没有出现。 [Spinning Up: SAC](https://spinningup.openai.com/en/latest/algorithms/sac.html)中也有强调，
> Unlike in TD3, the next-state actions used in the target come from the current policy instead of a target policy.

请问下，加上target network是为了得到更稳定的actor吗？

**Comment by Yonv1943** [comment]
Yes. 
1. SAC.2018 don't use act target network.
2. Act target network will bring a more stable training. (for SAC TD3 DDPG etc.)
3. If you have got a good set of hyperparameters, remove act target network will speed up traning slightly. (A good trick on paper, but a bad behavior for actual use)

So I decide to remove act target network in `AgentSAC`, and keep it in `AgentModSAC`.  What is your opinion?

你是正确的，**SAC的作者在2018年的论文中认为SAC不需要 target_network**，spinning up 忠实于原论文，也没有在act 上使用 target network。

然而，我自己在实现代码的时候进行了大量的实验，有 act_target_net 的 SAC、TD3、DDPG 的learning curve都要比没有这个结构的更稳定。我也发现，在超参数都很完美的情况下，去掉 act_target_net 能让训练变得更快！我猜测这可能是在论文里面刷分的一个技巧吧。

综上，出于「实用主义」的考虑，我还是在SAC中加上了 act_target_net 。毕竟我个人认为 robust 比起那些微不足道的性能提升要重要得多。然而，我又实现了 ModSAC（Modified SAC）。

因此，我决定，让这里实现的SAC算法忠于原论文，去掉act_target_net ，而我有什么想法可以直接加到 ModSAC里。请问你是否支持我这么做？我想知道你的看法。

**Comment by wsgdrfz** [comment]
Hi,

Thanks for your reply! In my view, yes, of course, that stability is the most important thing at the beginning, especially when trying new ideas. But later, I would like to fine-tune all the parameters including whether use the target network for the actor.

Now I met a problem that I tried both `AgentSAC` and `AgentModSAC ` for one self-designed environment, that is, tracking some paths with square errors as the cost function in each step. It works well using `AgentSAC` but the actor is mainly not updating at all with `AgentModSAC` (loss is always 0). I not familiar with the TTUR methods, can you give me some tips?
-----------------------------------------------------------------------------------------
以下是机翻：

你好，

感谢您的回复！我认为，是的，当然，起步时稳定是最重要的，特别是在尝试新想法时。但是稍后，我想微调所有参数，包括是否为Actor使用目标网络。

现在，我遇到了一个问题，我在一个自行设计的环境中尝试了“ AgentSAC”和“ AgentModSAC”，任务是追踪路径 并且使用误差的平方作为成本函数。使用`AgentSAC`可以很好地工作，但是actor主要不会使用`AgentModSAC`进行更新。我不熟悉TTUR方法，可以给我一些提示吗？

**Comment by Yonv1943** [comment]
The structure of AgentSAC and AgentModSAC are highly similar, only the update_net() function is different, and it is the same in other places, so there is no difference in use. It is not necessary to be familiar with the TTUR method to use the ModSAC algorithm. (TTUR's idea: updating the Actor network will increase the error of the Critic network. In order to reduce the error of the Critic, we choose to update the Critic several times, and even choose a smaller learning rate for the Actor network.)
I am currently fine-tuning the network structure of the ModSAC algorithm, which will be updated tonight.

The above content is the result of Google translation.
---

AgentSAC 和 AgentModSAC 的结构高度相似，只有 update_net() 这个函数不同，其他地方相同，使用起来没有区别。使用ModSAC算法不需要熟悉TTUR方法。（TTUR的思想： 更新Actor网络会增加Critic网络的误差，为了减小Critic的误差，我们选择多更新几次Critic，甚至给Actor网络选择更小的学习率。）
我最近正在微调ModSAC算法的网络结构，更新ModSAC 后，我会把这一句话改掉。

**Comment by wsgdrfz** [comment]
Yes, the difference between the two agents is when to update the actor. But I met the problem that the actor only updated once at the beginning of training. So just want to know how to decide when to update the actor? Is there any possibility that the algorithm never updates the actor?


是的，两个Actor之间的区别在于何时更新Actor的权重。但是我遇到了一个问题，那就是Actor仅仅在训练开始时更新了一次。所以只想知道如何决定何时更新Actor？TTUR算法是否有可能永远不会更新actor？

**Comment by Yonv1943** [comment]
*The above content is the result of Google translation.*

---
我以前是做对抗网络（GANs）的，TTUR是现在对抗网络中得到使用的思想。我发现TTUR适用于（对抗网络与策略梯度这类）双层优化问题（bi-level optimization）。详见： Connecting Generative Adversarial Networks and Actor-Critic Methods. 2017.
我已经在两年前（2019）的实验中尝试了各种与TTUR相关的方案。目前ModSAC使用的TTUR方案是我自以为最好的：

TTUR认为：若critic的误差过大，那么我们就不能用 critic提供的梯度去更新actor，或者应该用更小的步长更新actor。于是有TTUR的方案：确定三个超参数，让学习率 lr1 < lr2, 选择一个正整数n（一般是 2到4）
1. 用学习率 lr1 更新一次actor
2. 用学习率 lr2 更新n次 critic
3. 重复步骤 1 与 步骤 2

但是上面这个TTUR方案不够灵活，任何一个刚入门深度学习都能独立地想出来。于是我有一个进阶版的TTUR：
我认为可以使用  loss of critic 来衡量critic提供的梯度的误差，loss of critic 的值越大，证明 critic 提供的梯度越不可信。于是我用先用 soft update 得到一个变化幅度小的 loss of critic `obj_c`，接着我把 `obj_c` ∈(0, +inf) 用 `exp(obj_c **2)` 映射到 `(0, 1)`，得到`reliable_lamdba` 用来表示信任程度（0% ~ 100%）：
```
# in agent.py,  class AgentModSAC, def update_net()

self.obj_c = 0.995 * self.obj_c + 0.05 * obj_critic.item()
reliable_lambda = np.exp(-self.obj_c ** 2) 
``` 
有了这个`reliable_lamdba` 后，我就使用这个数值来自动调整 TTUR 的三个超参数。我直接让 learning rate of actor 乘上这个 `reliable_lamdba`。接着我让actor 与 critic 的更新比率在 1.0 到 2.0 之间调整，避免只更新critic 而没有更新 actor 的糟糕情况出现。而且下面这种比较复杂的写法可以让 actor 的更新尽可能地均匀。
```
# in agent.py,  class AgentModSAC, def update_net()

self.cri_optimizer.zero_grad()
obj_critic.backward()
self.cri_optimizer.step()
self.soft_update(self.cri_target, self.cri)

'''objective of actor using reliable_lambda and TTUR (Two Time-scales Update Rule)'''
reliable_lambda = np.exp(-self.obj_c ** 2)  # for reliable_lambda
if_update_a = (update_a / update_c) < (1 / (2 - reliable_lambda))
if if_update_a:  # auto TTUR
    update_a += 1

    q_value_pg = torch.min(*self.cri_target.get_q1_q2(state, a_noise_pg))
    obj_actor = -(q_value_pg + logprob * alpha.detach()).mean()
    obj_actor = obj_actor * max(0.1, reliable_lambda)  # todo ceta0

    self.act_optimizer.zero_grad()
    obj_actor.backward()
    self.act_optimizer.step()
    self.soft_update(self.act_target, self.act)
```

问： how to decide when to update the actor?
答： `if_update_a = (update_a / update_c) < (1 / (2 - reliable_lambda))`   时刻统计并计算出  actor 的更新次数 与 ciritc 更新次数的比例，让这个比例维持在  1.0 到 2.0 之间 （具体数值会随着 `reliable_lambda` 的变化而调整）。

问：Is there any possibility that the algorithm never updates the actor?
答：我们在设计的时候，已经避免了 never updates the actor 这件事情发生。不过，当 loss of critic 突然增大的时候，我们的 Modified TTUR 会得到一个 非常接近于0 的 `reliable_lambda`， learning rate of actor 乘以这个数字后也会变得非常小，相当于完全停止对 Actor network 的更新。不过这种极端状况的持续时间很短：因为Critic可以很快地评估一个“几乎停止更新的 Actor”，接着  loss of critic 也会回落到一个正常的大小，`reliable_lambda` 也会上升到一个正常的大小，接着 Actor 又能回复正常的更新了。

下面这幅图是 InterSAC（它也用了和ModSAC相同的 Modified TTUR方案，只是我刚好找不到 ModSAC 的结果图片了），可以看到：

![BipdealWalkerHardCore_313score](https://github.com/AI4Finance-LLC/ElegantRL/blob/master/figs/BipdealWalkerHardCore_313score.png)

**Comment by wsgdrfz** [comment]
Thanks. It solves my doubts! 

PS 上条回复有格式错误，修改后可以关闭这个issue了
